crates/factor-key-value/tests/test.rs

```diff
@@ -0,0 +1,214 @@
+    // String::new("foo").as
+    assert_eq!(
+        labels,
+        state.key_value.allowed_stores().iter().collect::<Vec<_>>()
+    );
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn overridden_default_key_value_works() -> anyhow::Result<()> {
+    let runtime_config = toml::toml! {
+        [key_value_store.default]
+        type = "redis"
+        url = "redis://localhost:6379"
+    };
+    run_test_with_config_and_stores_for_label(
+        Some(runtime_config),
+        vec![RedisKeyValueStore],
+        vec!["default"],
+    )
+    .await
+}
+
+#[tokio::test]
+async fn custom_spin_key_value_works() -> anyhow::Result<()> {
+    let runtime_config = toml::toml! {
+        [key_value_store.custom]
+        type = "spin"
+    };
+    run_test_with_config_and_stores_for_label(
+        Some(runtime_config),
+        vec![SpinKeyValueStore::new(None)?],
+        vec!["custom"],
+    )
+    .await
+}
+
+#[tokio::test]
+async fn custom_spin_key_value_works_with_absolute_path() -> anyhow::Result<()> {
+    let tmp_dir = tempdir::TempDir::new("example")?;
+    let path = tmp_dir.path().join("custom.db");
+    let path_str = path.to_str().unwrap();
+    let runtime_config = toml::toml! {
+        [key_value_store.custom]
+        type = "spin"
+        path = path_str
+    };
+    run_test_with_config_and_stores_for_label(
+        Some(runtime_config),
+        vec![SpinKeyValueStore::new(None)?],
+        vec!["custom"],
+    )
+    .await?;
+    assert!(tmp_dir.path().exists());
+    Ok(())
+}
+
+#[tokio::test]
+async fn custom_spin_key_value_works_with_relative_path() -> anyhow::Result<()> {
+    let tmp_dir = tempdir::TempDir::new("example")?;
+    let path = tmp_dir.path().to_owned();
+    let runtime_config = toml::toml! {
+        [key_value_store.custom]
+        type = "spin"
+        path = "custom.db"
+    };
+    run_test_with_config_and_stores_for_label(
+        Some(runtime_config),
+        vec![SpinKeyValueStore::new(Some(path))?],
+        vec!["custom"],
+    )
+    .await?;
+    assert!(tmp_dir.path().exists());
+    Ok(())
+}
+
+#[tokio::test]
+async fn custom_redis_key_value_works() -> anyhow::Result<()> {
+    let runtime_config = toml::toml! {
+        [key_value_store.custom]
+        type = "redis"
+        url = "redis://localhost:6379"
+    };
+    run_test_with_config_and_stores_for_label(
+        Some(runtime_config),
+        vec![RedisKeyValueStore],
+        vec!["custom"],
+    )
+    .await
+}
+
+#[tokio::test]
+async fn misconfigured_spin_key_value_fails() -> anyhow::Result<()> {
+    let runtime_config = toml::toml! {
+        [key_value_store.custom]
+        type = "spin"
+        path = "/$$&/bad/path/foo.db"
+    };
+    assert!(run_test_with_config_and_stores_for_label(
+        Some(runtime_config),
+        vec![SpinKeyValueStore::new(None)?],
+        vec!["custom"]
+    )
+    .await
+    .is_err());
+    Ok(())
+}
+
+#[tokio::test]
+async fn multiple_custom_key_value_fails() -> anyhow::Result<()> {
```

Comment:
I guess this doesn't actually fail? It seems like we just ignore the second configuration. Is that what we want?

Reply:
This test led me to update the impl to check for existence of a store manager for a label before adding the label and manager. I updated this test name to be `multiple_custom_key_value_uses_first_store`. We could also error here